src/ppx/bucklescript/ppx.ml

```diff
@@ -1,10 +1 @@
-let () =
-  let argc = Array.length Sys.argv in
-  let argv =
-    let output_file = Sys.argv.(argc - 1) in
-    Array.(append
-      (sub Sys.argv 0 (argc - 1))
-      [|"-o"; output_file; "--dump-ast"; "--conditional"|])
-  in
-
-  Migrate_parsetree.Driver.run_main ~argv ()
+let () = Ppxlib.Driver.run_as_ppx_rewriter ()
```

Comment:
This appears to lose the effect of the `--conditional` flag. In your local testing with BuckleScript, are you able to compile the same project xwithout `BISECT_ENABLE=yes` such that instrumentation is disabled, and with `BISECT_ENABLE=yes` such that instrumentation is enabled?

Reply:
The bucklescript test always yields `Coverage: 1/1 (100.00%)`, both with and without `BISECT_ENABLE=yes` and both on my branch and on master. I think the reason for that is that the `conditional` reference is looked at when registering the PPX. At that point it always contains `false`. On master, it gets set to `true` when executing the driver, but at that point the PPX is already registered. But probably I'm missing something. Does using the `--conditional` flag usually have an impact? 

I could set the conditional reference to true here (`Bisect_ppx.Register.conditional := true`), but I have the problem that I'm not sure if that would make a difference. Do you think it would? Do you want me to do that?

Why are there two things in the context of enabling or disabling bisect (the `-conditional` flag and the `BISECT_ENABLE` environment variable)? Does dune set any of them? Or ocamlbuild?

Reply:
> The bucklescript test always yields `Coverage: 1/1 (100.00%)`, both with and without `BISECT_ENABLE=yes` and both on my branch and on master.

I can't reproduce that on `master`. Removing `BISECT_ENABLE=yes` from `test/bucklescript/package.json` and doing a clean run of the tests, so that there are no Bisect artifacts left from previous runs with `BISECT_ENABLE=yes`, I get:

```
./node_modules/.bin/bisect-ppx-report html
Error: no coverage files given on command line or found
```



> Does using the `--conditional` flag usually have an impact?
> 
> I could set the conditional reference to true here (`Bisect_ppx.Register.conditional := true`), but I have the problem that I'm not sure if that would make a difference. Do you think it would? Do you want me to do that?

Yes, it has an impact; yes it would make a difference. As for what to do, I originally found it convenient to pass the argument the same way a user would have before the recent improvement to the Dune integration, mainly to ensure I trigger exactly the same logic. I still don't know yet if ppxlib treats arguments the same way. If so, then I suggest passing the argument. If not, you can try setting the reference directly, or using any other reasonable way.

> Why are there two things in the context of enabling or disabling bisect (the `-conditional` flag and the `BISECT_ENABLE` environment variable)? Does dune set any of them? Or ocamlbuild?

For backward compatibility. Originally, Bisect ran unconditionally if it was included in the PPX list at all, and this was fine because it was easy to toggle it from Ocamlbuild. Once Dune came out, we added `--conditional` to make Bisect off by default, since there was no easy way to remove it from the preprocessor list with Dune (short of writing `dune` files in OCaml syntax).

Reply:
Ok, thanks for the explanation! I've set the conditional reference now. Could you maybe check if that does what we want? I've tried and it keeps on doing the same thing for me, but since for me that's also the case on master I must be doing something wrong running the bucklescript test.

At the moment, in Ppxlib it isn't possible to pass arguments to the function that creates the driver. And if it's not necessary, we'd like to avoid that. So I'm trying to find out if it's necessary.

Reply:
Can you state exactly what you tried?

For reference:

1. Remove `BISECT_ENABLE=yes` from `package.json` in `test/bucklescript`.
2. Either run `make clean` in `test/bucklescript` (better) or do `rm -f *.coverage` in that directory.

Re-run the BuckleScript test.

Reply:
What I'm doing is the following:

1. In `src/npm-install.sh`, I've prefixed the path to the driver (lines 8 and 10) with `../../..`. That is because -due to something I'm doing wrong, but I don't know what- the script is in `test/bucklescript/package` at that point and there it can't find the binary. This might be the problem.
2. I've removed `BISECT_ENABLE=yes` from package.json in `test/bucklescript`.
3. From the root of my project I run `make -C test/bucklescript/ full-test` (which should include `make clean`).

Reply:
But the way I'm running the test it does work to some extend, since it used to run but fail before fixing Ppxlib and now it passes.

Reply:
I've continued the discussion [here](https://github.com/aantron/bisect_ppx/pull/327#pullrequestreview-583006193), since that line looks like the source of the underlying problem with `--conditional` being meaningless in the new code, and there is also a way to test for this using Dune rather than BuckleScript.